### ISSUE TITLE
User profile: postal addresses

### ISSUE THREAD

**Issue by mkalininait** [rails-girls-summer-of-code/rgsoc-teams#868]
RGSoC is an international program, and it's sometimes hard to parse postal addresses of people from other countries.

**Suggestion:**

User profile, edit form: instead of one input field "Postal address" implement several (* are required):
 - Recipient's name*
 - Company
 - Address Line 1*
 - Address Line 2
 - City*
 - State/Province/Other 
 - Postal code*
 - Country*
 - Phone number

**Comment by nerdyvaishali** [comment]
I think Phone number should be a mandatory field as the whole point of asking this infromation is to send packages. For the privacy issues, may be this fields should be only visible to organizers [or a person who handles swag packaging].

**Comment by carpodaster** [comment]
> RGSoC is an international program, and it's sometimes hard to parse postal addresses of people from other countries.

That's why I would think that a multi-line textarea is better than having individual fields. Every country formats postal address differently. If a user puts their address exactly the way how it must appear on the envelope, shouldn't we be fine?

**Comment by alicetragedy** [comment]
@carpodaster yeah, I think that was the original thought behind the text area.

The issue with the packages is that we need to fill out custom forms for each of the non-EU packages, and it some cases it was difficult to fill out from a text block in a text area. We were unsure if “house behind the only bus stop in town” is an acceptable entry for “Street” 😂  (I'm obviously exaggerating, but some of the addresses were very cryptic and it was hard to tell if this was a regular local address or not, so we had to look that up manually and/or ask people, which was a bit of a time suck).

Also: I'm not sure if this has anything to do with the actual ticket, but last year for swag we looked into partly automatising the process, i.e. exporting the data, uploading it to the platform of the chosen shipping service (e.g. postal service, dhl, fedex, hermes), and having the packages be picked up rather than doing everything manually. It could be that these services ask for the data to be formatted in a certain way.

**Comment by shebesabrina** [comment]
Hey @mkalininait, I wanted to see if I could help with this issue. It seems you're needing some help with mandatory fields including phone numbers, and addresses. As well as making sure you have everything needed for shipment anywhere in the world. Mind if I try giving this a shot?

**Comment by anasofiapinho** [comment]
Hello 👋 I have not been in charge of sending the packages so far, but I have sent some and I was in charge of collecting the addresses back in 2016. I think @alicetragedy is the best person to speak about this though, but as I am seeing that you are considering the removal of some fields, I just wanted to give my two cents. — and I am sorry if I am just raising problems here. If you feel that they aren't important or that they will block the resolution of this issue, please ignore.

As @nerdyvaishali mentioned, phone number are usually mandatory because in some countries the shipping companies call the receiver before delivery. So I suggest that you don't remove that field.

Regarding Address Line 2, yes @klappradla I have had issues because the shipping company didn't receive address line 2, but usually they also don't receive the full text that I put on address line 1. If you live in Portugal, sometimes address names are so long that we have to use 2 lines. I.e.: "Av. Dr. Domingos Caetano de Sousa _House#_ _Floor_ ", or we need to add the small locality besides the street name and number. This is why having the phone number is important, so that the shipping companies can call us and having the tracking codes is even better because I can detect when the packages are stuck somewhere.
I don't know if the fields here will have a lot of characters or if you prefer to keep address line 2, I just wanted to let you know this. I also took a look at this year's team's addresses and they are very different. Some have up to 6 elements separated by comas while others have 2 lines.

I have another concern though: if we are telling the students to add their addresses in this order, but the elements in the addresses in their country have different orders, would it cause any issues when the packages got to the country? Imagine that a city had a similar name to a locality and the order was different, so the postman/postwoman sent it to a different city.

**Comment by shebesabrina** [comment]
Hi all 👻!
I've started the process of submitting a pull request so there are multiple fields for an address, here's a screenshot of what the edit and user show pages look like:

<img width="1176" alt="screen shot 2018-09-05 at 8 01 34 am" src="https://user-images.githubusercontent.com/18236509/45107566-3953bd00-b0f7-11e8-947a-2ed7211f5fd9.png">

<img width="1315" alt="screen shot 2018-09-05 at 8 01 19 am" src="https://user-images.githubusercontent.com/18236509/45107724-bd0da980-b0f7-11e8-93eb-a5f6e06665ba.png">

It seems you would also like to include a second address line under "street" (I'm thinking of labeling it as street_1 and street_2 in the table and view. This I think I can do 🎉). My one clarifying question is if there needs to be a mandatory field for the user/student's phone number or if you need the RGSoC contact number included in every postal_address so the post office can contact you for additional clarification on the address.

**Comment by klappradla** [comment]
Thanx @anasofiapinho! Makes all perfect sense!

And @shebesabrina: wow, that was quick! Thanx for your help on this and for already submitting a PR. Based on the discussion here, we can just iterate on it until we got what's most helpful for orga users.

**Comment by alicetragedy** [comment]
Hiya! thanks so much for this @shebesabrina 🙌 ✨ 
To be perfectly honest I feel similarly to @klappradla on the topic of the phone number. I've provided my personal phone number in several instances when ordering or shipping things, and that number was never used when problems in delivery were encountered (sometimes leading to the package just being sent back without me ever being told that they tried to deliver it), so by now I feel like providing this information is kind of useless 😅 
We currently don't have a fixed phone number for RGSoC, but we are working on that. So perhaps for the time being it would make sense to not have a phone number field at all :) The next time we ship packages, we can write our own number ourselves if necessary, and it doesn't need to go through the app.
If we *do* have a phone number field (optional), perhaps it would be good to leave a “hint” as to why this is needed — to make clear that it's not necessary. I personally try not to fill out too private information when filling out forms if it's not necessary but perhaps not everyone has the same approach to privacy.


As for the “street” fields: since not all non-western addresses will have the concept of streets in the address, perhaps something like address line 1 and address line 2 would make more sense. whether that's only the label or the database field I'd leave up to you, though perhaps for consistency's sake it would be good to have the database column have the same name as the label, so that this doesn't get accidentally changed in the future.

**Comment by lilwillifo** [comment]
Hey @klappradla -- It looks like the related PR never was merged. I'm happy to take a look at those notes and work on this issue. I actually know @shebesabrina who said she'd be fine with me picking up where she left off! Small open source world 😆Let me know if there's any new context I'll need around this. Thank you!

**Comment by klappradla** [comment]
Hey @lilwillifo 👋 

Welcome!

I don't think or at least don't remember there being any more discussion on the topic. So anything written here and in the PR should be all you need. Just ping me if something is not clear!
